File: scala-spring/commandside-web-transactions/src/main/scala/net/chrisrichardson/eventstore/examples/bank/web/transactions/controllers/MoneyTransferController.scala
package net.chrisrichardson.eventstore.examples.bank.web.transactions.controllers

import net.chrisrichardson.eventstore.EntityId
import net.chrisrichardson.eventstore.EventStore
import net.chrisrichardson.eventstore.examples.bank.backend.common.transactions.TransferDetails
import net.chrisrichardson.eventstore.examples.bank.transactions.{MoneyTransferService, MoneyTransfer}
import net.chrisrichardson.eventstore.examples.bank.web.util.WebUtil
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation._

import scala.concurrent.ExecutionContext.Implicits.global

@RestController
class MoneyTransferController @Autowired()(moneyTransferService : MoneyTransferService,
                                        eventStore : EventStore) {

  @RequestMapping(value=Array("/transfers"), method = Array(RequestMethod.POST))
  def create(@RequestBody transferDetails : TransferDetails) = WebUtil.toDeferredResult {
    for (transaction <- moneyTransferService.transferMoney(transferDetails))
      yield CreateMoneyTransferResponse(transaction.entityId.id)
  }

  @RequestMapping(value=Array("/transfers/{transferId}"), method = Array(RequestMethod.GET))
  def get(@PathVariable transferId : String) = {
    val f = eventStore.find[MoneyTransfer](EntityId(transferId))
    WebUtil.toDeferredResult(f map(transactionAggregate => GetMoneyTransferResponse(transferId, transactionAggregate.entity.state.getClass.getSimpleName)))
  }

}

File: java-spring/transactions-command-side-web/src/main/java/net/chrisrichardson/eventstore/javaexamples/banking/web/commandside/transactions/MoneyTransferController.java
package net.chrisrichardson.eventstore.javaexamples.banking.web.commandside.transactions;


import net.chrisrichardson.eventstore.javaexamples.banking.backend.commandside.transactions.MoneyTransferService;
import net.chrisrichardson.eventstore.javaexamples.banking.backend.common.transactions.TransferDetails;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

import java.util.concurrent.CompletableFuture;

import java.util.Date;

@RestController
@RequestMapping("/transfers")
public class MoneyTransferController {

  private final MoneyTransferService moneyTransferService;

  @Autowired
  public MoneyTransferController(MoneyTransferService moneyTransferService) {
    this.moneyTransferService = moneyTransferService;
  }

  @RequestMapping(method = RequestMethod.POST)
  public CompletableFuture<CreateMoneyTransferResponse> createMoneyTransfer(@Validated @RequestBody CreateMoneyTransferRequest request) {
    TransferDetails transferDetails = new TransferDetails(
            request.getFromAccountId(),
            request.getToAccountId(),
            request.getAmount(),
            new Date(),
            request.getDescription());
    return moneyTransferService.transferMoney(transferDetails)
            .thenApply(entityAndEventInfo -> new CreateMoneyTransferResponse(entityAndEventInfo.getEntityIdAndVersion().getEntityId()));
  }

}
